GowerMetric.py

```diff
@@ -0,0 +1,371 @@
+import os.path
+import timeit
+from pathlib import Path
+from typing import List, Optional
+
+import numpy as np
+import sklearn.utils.validation
+from scipy.spatial.distance import pdist
+from scipy.cluster.hierarchy import (
+    linkage,
+    cophenet,
+    fcluster,
+)
+from scipy.optimize import minimize
+from sklearn.metrics import (
+    silhouette_score,
+)
+from sklearn.preprocessing import OrdinalEncoder, StandardScaler
+
+from utils import DataType
+
+
+class GowerMetric:
+    def __init__(
+        self,
+        dtypes: np.array,
+        ratio_scale_normalization="range",
+        weights=None,
+        precomputed_weights_file=None,
+    ):
+        self.dtypes = dtypes  # initialize with np.array of column data types
+        self.weights = weights
+        self.precomputed_weights_file = precomputed_weights_file
+        self.ranges_: List  # values of ranges in .ratio_scale() (iqr or traditional range)
+        self.h_: List  # h values in .ratio_scale()
+        self.n_features_in_: int
+        self.ratio_scale_normalization: str = ratio_scale_normalization
```

Comment:
We should do parameter validation here for `ratio_scale_normalization`. Possible values are: "range", "iqr", "kde"